### PR TITLE
Config parser

### PR DESCRIPTION
**Pull request by PNieck**
## Description

This PR uses `parser` library to get PIDs and other configuration information from file.

## Motivation and Context

No code repetition. Two different modules don't do the same thing.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (list targets here). `armv7a9-zynq7000-qemu`

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
https://github.com/phoenix-pilot/phoenix-pilot-core/pull/175
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Review by HBuczynski — CHANGES_REQUESTED** [review]
The code in `control` looks much better after your changes.

General remark. Please provide the same API for all functions in `config.h`. If any changes in the future happens, handling it will be much easier.

A function in `config.h' should focus only on parsing dedicated structure, not the whole file. Making association with other headers within a file is not a good approach. The changes in parser library are necessary  to handle such situation.

File: quadcontrol/config.c [hunk 1]
```diff
@@ -30,126 +29,162 @@ static struct {
...
 		return -1;
 	}
 
-	mode->type = flight_pos;
+	return 0;
+}
+
+
+/* Doubles the `res[parserType]` buffer if necessary  */
+static int config_handleSpace(int parserType, size_t dataTypeSize)
```
Comment (HBuczynski) [inline quadcontrol/config.c:92]: `config_reallocStructs` or `config_reallocData` is more suitable.

File: quadcontrol/config.c [hunk 2]
```diff
@@ -30,126 +29,162 @@ static struct {
...
+		res[parserType].sz *= reallocRate;
+	}
 
 	return 0;
 }
 
 
-static int config_hoverParse(const hmap_t *h, flight_mode_t *mode)
+/* If necessary changes the size of res[parserType].data` so that it matches number of elements in this table */
+static int config_adjustSpace(int parserType, size_t dataTypeSize)
```
Comment (HBuczynski) [inline quadcontrol/config.c:112]: `config_trimUnusedStructs` or `config_trimUnusedData`

File: quadcontrol/config.c [hunk 3]
```diff
@@ -277,3 +297,233 @@ int config_scenarioRead(const char *path, flight_mode_t **scenario, size_t *sz)
...
+	int err;
+	parser_t *p;
+	static const unsigned int initSz = 4;
+
+	if (path == NULL || pids == NULL || sz == NULL) {
+		fprintf(stderr, "config: invalid arguments\n");
+		return -1;
+	}
+
+	p = parser_alloc(1, 7);
```
Comment (HBuczynski) [inline quadcontrol/config.c:343]: Somme comment about `(1,7) magic numbers would be a good information`

File: quadcontrol/config.c [hunk 4]
```diff
@@ -277,3 +297,233 @@ int config_scenarioRead(const char *path, flight_mode_t **scenario, size_t *sz)
...
+{
+	int err;
+	parser_t *p;
+	static const unsigned int initSz = 1;
+
+	if (path == NULL || throttle == NULL || sz == NULL) {
+		fprintf(stderr, "config: invalid arguments\n");
+		return -1;
+	}
+
```
Comment (HBuczynski) [inline quadcontrol/config.c:417]: You can apply this for other functions:

```suggestion
*throttle = NULL;
*sz = 0;
```

File: quadcontrol/config.h [hunk 5]
```diff
@@ -15,10 +15,23 @@
...
 
 #include "control.h"
+#include "pid.h"
 
 
 /* Parses file from `path` with saved mission scenario. Returns array `scenario` with length equal to `sz` */
 extern int config_scenarioRead(const char *path, flight_mode_t **scenario, size_t *sz);
 
 
+/* Parser all PIDs from file defined by `path`. Returns array `pids` with length equal to `sz` */
```
Comment (HBuczynski) [inline quadcontrol/config.h:25]: Nope.
It returns 0 oraz -1.

File: quadcontrol/control.c [hunk 6]
```diff
@@ -770,14 +549,60 @@ static int quad_init(void)
...
+
+	quad_common.targetAtt = *attTmp;
+	free(attTmp);
+#endif
+
 	if (config_scenarioRead(PATH_SCENARIO_CONFIG, &quad_common.scenario, &quad_common.scenarioSz) != 0) {
 		resourceDestroy(quad_common.rcbusLock);
+		free(quad_common.pids);
 		return -1;
 	}
```
Comment (HBuczynski) [inline quadcontrol/control.c:607]: `quad_init` function has become too big.
 I would suggest to put parsing logic in a `quad_config` functions.

**Review by HBuczynski — CHANGES_REQUESTED** [review]
General code looks fine.
I have pointed out some small remarks.

**Review by niewim19 — DISMISSED** [review]
LGTM

**Comment by PNieck** [comment]
I have changed the name of define form `PATH_QUAD_CONFIG` to `PATH_QUAD_CONFIG`

**Review by HBuczynski — APPROVED** [review]
LGTM, nice improvement in quad-control unit.

**Pull request merged**